src/cursor/cur_hs.c

```diff
@@ -88,7 +107,14 @@ __curhs_file_cursor_next(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
 {
     WT_DECL_RET;
 
-    WT_WITH_TXN_ISOLATION(session, WT_ISO_READ_UNCOMMITTED, ret = cursor->next(cursor));
+    /*
+     * Do not relax the isolation level when reading from a checkpoint. We want to see what was
+     * committed as of the checkpoint and not see anything committed later.
+     */
+    if (WT_READING_CHECKPOINT(session))
+        ret = cursor->next(cursor);
```

Comment:
I think so -- I put that in because it was failing. But it was a while back and I don't remember the exact details (one of the problems with long-running branches like this...)

I think the following scenario requires it: (1) generate an inconsistent checkpoint with a torn transaction in it, so we need to look at the history store checkpoint to get the latest committed value as of the checkpoint; (2) write some extra unstable stuff to the history store before it gets checkpointed (as seen in WT-7657); (3) read the checkpoint with no timestamp. Then we go to the history store checkpoint and read it with no timestamp, and that'll return the latest thing even though it's unstable and not supposed to be visible in the checkpoint, instead of the value from before the original torn transaction.

Since there's no RTS when reading from a checkpoint we have to be careful to produce the same results by using the checkpoint's snapshot.

Where's the other check you're thinking of? Maybe it should be patched instead.

Reply:
As per my understanding, the scenario that you described cannot cause any problems due to the checkpoint snapshot that is used to compare while retrieving the data from the history store. 

The newly inserted records into the history store even with the commit timestamp as a stable timestamp, but the transaction id that performed these updates is not visible according to the checkpoint snapshot. In the functions `__curhs_prev_visible` or `__curhs_next_visible` already validates the returned value with the session transaction visibility. I think we need to make sure that WT_CURSTD_HS_READ_COMMITTED or WT_CURSTD_HS_READ_ALL flags are not set as part of the checkpoint history store cursor. The existing code takes care of the checkpoint cursor visibility. 

Please let me know if I miss anything?

Reply:
It's more likely that I missed something. Or that I needed it because it was before the logic for getting a matching history store checkpoint worked. I'll look into it.

Reply:
(Have not forgotten this, but haven't gotten to it yet)

Reply:
Ok, as far as I can tell you're right. I don't see any way that `WT_CURSTD_HS_READ_COMMITTED` or `WT_CURSTD_HS_READ_ALL` can get set inappropriately, so the checks in `__curhs_prev_visible` and `__curhs_next_visible ought` to be sufficient. There's a global visibility check in there, but it's already been patched to handle the checkpoint cursor case.

I can't remember what I was seeing when I made the change. It was in early January, so it's been two and a half months. :-|  According to the unsquashed commit messages the problem involved finding pieces of a transaction that wasn't visible in the checkpoint but was fully committed later, which suggests that the history store checkpoint didn't match the data store checkpoint. This was very early on and before I'd even noticed that issue, AFAICR, and while I was still working with the simpler tests. So I think unless something breaks we can assume it's a nonissue.

I'm going to disable this check and run tests to make sure nothing breaks, but assuming nothing unexpected happens I'll remove it.

Reply:
Removed it in [d1c9081](https://github.com/wiredtiger/wiredtiger/pull/7661/commits/d1c90812c87ae43f4e3cbce2ac32ed59e30f4b8d)